hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/FileSystem.java

```diff
@@ -2413,8 +2413,13 @@ private void handleFileStat(LocatedFileStatus stat) throws IOException {
         if (stat.isFile()) { // file
           curFile = stat;
         } else if (recursive) { // directory
-          itors.push(curItor);
-          curItor = listLocatedStatus(stat.getPath());
+          try {
+            RemoteIterator<LocatedFileStatus> newDirItor = listLocatedStatus(stat.getPath());
```

Comment:
do you need to handle the condition where the dir has been deleted and replaced with a file? as it is the other concurrency failure, isn't it?

Reply:
That wasn't an issue. If you do ``listLocatedStatus(stat.getPath());`` and the path is a file, it will still return an iterator with just one entry which is that file, and it will again loop on that iterator find the first entry as file and add it to the list, So, that won't break but I have added a mocked test as well for this case.